scipy/_lib/_array_api.py

```diff
@@ -217,26 +224,43 @@ def _assert_matching_namespace(actual, desired):
     for arr in actual:
         arr_space = array_namespace(arr)
         assert_(arr_space == desired_space,
-                "Namespaces do not match: "
-                f"{arr_space.__name__} and {desired_space.__name__}")
-        if is_numpy(arr_space):
-            _check_scalar(actual, desired)
-
-
-def _check_scalar(actual, desired):
-    for arr in actual:
-        if np.isscalar(desired):
-            assert np.isscalar(arr), "Desired a scalar, but ndarray given."
-        else:
-            assert not np.isscalar(arr), "Desired an ndarray, but scalar given."
-
-
-def xp_assert_equal(actual, desired, check_xp=True, check_dtype=True, 
+                "Namespaces do not match.\n"
+                f"Actual: {arr_space.__name__}\n"
+                f"Desired: {desired_space.__name__}")
+
+
+def _check_scalar(actual, desired, xp):
+    # Shape check alone is sufficient unless desired.shape == (). Also,
+    # only NumPy and CuPy distinguish between scalars and arrays, AFAICT.
+    if desired.shape != () or not (is_numpy(xp) or is_cupy(xp)):
+        return
+    # We want to follow the conventions of the `xp` library. Libraries like
+    # NumPy, for which `np.asarray(0)[()]` returns a scalar, tend to return
+    # a scalar even when a 0D array might be more appropriate:
+    # import numpy as np
+    # np.mean([1, 2, 3])  # scalar, not 0d array
+    # np.asarray(0)*2  # scalar, not 0d array
+    # np.sin(np.asarray(0))  # scalar, not 0d array
+    # Libraries like CuPy, for which `cp.asarray(0)[()]` returns a 0D array,
+    # tend to return a 0D array in scenarios like those above.
+    # Therefore, regardless of whether the developer provides a scalar or 0D
+    # array for `desired`, we would typically want the type of `actual` to be
+    # the type of `desired[()]`. If the developer wants to override this
+    # behavior, they can set `check_shape=False`.
```

Comment:
Does this make sense?

Reply:
Yes I think so, although modulo CuPy https://github.com/scipy/scipy/pull/19186#discussion_r1320502626?